csharp/src/SeedLang/Runtime/NativeFunctions.cs

```diff
@@ -17,73 +17,115 @@
 
 namespace SeedLang.Runtime {
   using NativeFunction = HeapObject.NativeFunction;
-  using Range = HeapObject.Range;
 
   // The static class to define all the build-in native functions.
   internal static class NativeFunctions {
     public const string Append = "append";
+    public const string Eval = "eval";
     public const string Len = "len";
     public const string List = "list";
+    public const string Print = "print";
     public const string Range = "range";
 
     public static NativeFunction[] Funcs = new NativeFunction[] {
-      // Appends a value to a list. The first argument is the list, the second argument is the
-      // value to be appended to the list.
-      new NativeFunction(Append, (Value[] args, int offset, int length) => {
-        if (length != 2) {
-          throw new DiagnosticException(SystemReporters.SeedRuntime, Severity.Fatal, "", null,
-                                        Message.RuntimeErrorIncorrectArgsCount);
-        }
-        Value value = args[offset];
-        if (value.IsList) {
-          List<Value> list = value.AsList();
-          list.Add(args[offset + 1]);
-        }
-        return new Value();
-      }),
+      new NativeFunction(Append, AppendFunc),
+      new NativeFunction(Eval, EvalFunc),
+      new NativeFunction(Len, LenFunc),
+      new NativeFunction(List, ListFunc),
+      new NativeFunction(Print, PrintFunc),
+      new NativeFunction(Range, RangeFunc),
+    };
 
-      new NativeFunction(Len, (Value[] args, int offset, int length) => {
-        if (length != 1) {
-          throw new DiagnosticException(SystemReporters.SeedRuntime, Severity.Fatal, "", null,
-                                        Message.RuntimeErrorIncorrectArgsCount);
-        }
-        return new Value(args[offset].Length);
-      }),
+    // Appends a value to a list. The first argument is the list, the second argument is the
+    // value to be appended to the list.
+    private static Value AppendFunc(Value[] args, int offset, int length, VisualizerCenter vc,
+                                    Range range) {
+      if (length != 2) {
+        throw new DiagnosticException(SystemReporters.SeedRuntime, Severity.Fatal, "", null,
+                                      Message.RuntimeErrorIncorrectArgsCount);
+      }
+      Value value = args[offset];
+      if (value.IsList) {
+        List<Value> list = value.AsList();
+        list.Add(args[offset + 1]);
+      }
+      return new Value();
+    }
 
-      // Creates an empty list if the length of arguments is empty, and a list if the argument is a
-      // subscriptable value.
-      new NativeFunction(List, (Value[] args, int offset, int length) => {
-        if (length < 0 || length > 1) {
+    // Evaluates a value when it's not none. It's used by SeedCalc and interactive mode SeedX
+    // primarily.
+    private static Value EvalFunc(Value[] args, int offset, int length, VisualizerCenter vc,
```

Comment:
Python3's `eval` function accepts a string as its parameter. This is different from our current implementation, which actually accepts a runtime value/expression as its parameter.

```
python3
Python 3.9.10 (main, Jan 15 2022, 11:48:04) 
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> eval(3+4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: eval() arg 1 must be a string, bytes or code object
>>> eval("3+4")
7
```

```
dotnet run --project src/SeedLang.Shell
SeedLang.Shell 0.2.0-preview20220227142557
Copyright 2021-2022 The SeedV Lab.

>>> eval(3+4)
---------- Source ----------
1     eval(3+4)
---------- Run ----------
1     eval(3+4)
Eval: 7
```


Do we plan to make our `eval` function consistent with Python's design?

If the answer is yes, we should at least add a TODO comment here.

If it's not the case, what's the reason we need to expose this function to users? What can our users do with this function? I think in SeedLang.Shell we can implicitly call this function without letting our users know that, can't we?

And, if we really need an `eval` function that accepts a runtime value (expression) as its parameter,  can we name this built-in function slightly different from Python's? Probably we can name it `valeval` which means "value evaluation".

Reply:
This function is not intended to be `eval`. It used to handle the single expression statement, and output the value. I think only SeedCalc and SeedLan.Shell need it.
We need this function because compiler need place It in the byte code to print the value of expression statement. SeedLang.Shell doesn't know when to call it.
How about change it to a internal build-in function like `__evalval__`?

For the real `eval` function, we can decide if we need it later.

Reply:
`__evalval__` sounds perfect to me.

Reply:
Rename it `__printval__` eventually, because this function doesn't evaluate anything, it just print the string representative of a value if it's not none in the interactive mode.